### PR TITLE
ENH/BUG: Support string slicing with other expressions

### LINKED ISSUE #1626: ENH/BUG: Support string slicing with other expressions
```python
t = ibis.table([('a', 'string'), ('b', 'int64')])
expr = t.a[:t.b]
```

This currently fails with a `ValueError` because we're assuming slice inputs to `__getitem__` are integers.

### PR DESCRIPTION
**Pull request by cpcloud**
Closes #1626

### CODE REVIEW

File: ibis/expr/api.py [hunk 1]
```diff
@@ -2016,21 +2016,28 @@ def _string_dunder_contains(arg, substr):
...
+            if start is None:
+                start = 0
+            if start < 0:
+                raise ValueError('Negative slicing not yet supported')
 
-        if start < 0 or stop < 0:
-            raise ValueError('negative slicing not yet supported')
+        if not isinstance(stop, ir.Expr):
+            if stop is not None and stop < 0:
+                raise ValueError('negative slicing not yet supported')
```
Comment (kszucs) [inline ibis/expr/api.py]: Uppercase `n`

File: ibis/expr/api.py [hunk 2]
```diff
@@ -2016,21 +2016,28 @@ def _string_dunder_contains(arg, substr):
...
+            if stop is None:
+                stop = self.length()
```
Comment (kszucs) [inline ibis/expr/api.py]: Minor, just not consistent with the previous validation order.

File: ibis/expr/api.py [hunk 3]
```diff
@@ -2016,21 +2016,28 @@ def _string_dunder_contains(arg, substr):
...
 
         return self.substr(start, stop - start)
     elif isinstance(key, six.integer_types):
         return self.substr(key, 1)
-    else:
-        raise NotImplementedError(
-            'string __getitem__[{}]'.format(type(key).__name__)
-        )
+    raise NotImplementedError(
```
Comment (kszucs) [inline ibis/expr/api.py]: I regularly think about `else or not to else` :)

File: ibis/pandas/execution/strings.py [hunk 4]
```diff
@@ -26,16 +26,50 @@ def execute_string_length_series(op, data, **kwargs):
...
+            return None
+        return value[start:end]
+    return data.map(iterate)
+
+
+@execute_node.register(ops.Substring, pd.Series, pd.Series, pd.Series)
+def execute_string_substring_series_series(op, data, start, length, **kwargs):
+    end = start + length
+
+    def iterate(value,
```
Comment (kszucs) [inline ibis/pandas/execution/strings.py]: This "general" case wouldn't be enough?

Comment (cpcloud) [reply ibis/pandas/execution/strings.py]: I suppose we could farm out the other functions to this one.

File: ibis/tests/all/test_string.py [hunk 5]
```diff
@@ -150,6 +150,33 @@ def test_string_col_is_unicode(backend, alltypes, df):
             lambda t: t.date_string_col.str[-2:],
             id='right',
         ),
+        param(
+            lambda t: t.date_string_col[1:3],
+            lambda t: t.date_string_col.str[1:3],
+            id='slice'
+        ),
+        param(
+            lambda t: t.date_string_col[t.date_string_col.length() - 1:],
```
Comment (kszucs) [inline ibis/tests/all/test_string.py]: In a follow-up PR we could add support for translating negative indices to `col.length() + <negative index>`

File: ibis/pandas/execution/strings.py (hunk 4 above)
Comment (cpcloud) [reply ibis/pandas/execution/strings.py]: We definitely need to handle the integer, series and series, integer cases distinctly, but those two can be made to be instances of the general case.

File: ibis/expr/api.py (hunk 2 above)
Comment (cpcloud) [reply ibis/expr/api.py]: I don't follow, what was the previous validation order?

File: ibis/expr/api.py [hunk 6]
```diff
@@ -2016,21 +2016,28 @@ def _string_dunder_contains(arg, substr):
...
+
+        if step is not None and not isinstance(step, ir.Expr) and step != 1:
             raise ValueError('Step can only be 1')
 
-        start = start or 0
+        if not isinstance(start, ir.Expr):
+            if start is None:
+                start = 0
+            if start < 0:
+                raise ValueError('Negative slicing not yet supported')
```
Comment (kszucs) [inline ibis/expr/api.py]: Perhaps `self.length() + start`?

File: ibis/expr/api.py (hunk 2 above)
Comment (kszucs) [reply ibis/expr/api.py]: `is None` goes to the top + `elif`

File: ibis/tests/all/test_string.py (hunk 5 above)
Comment (cpcloud) [reply ibis/tests/all/test_string.py]: yes, though why cross out the follow up?

Comment (cpcloud) [reply ibis/tests/all/test_string.py]: i don't think that should happen in this PR

File: ibis/expr/api.py (hunk 3 above)
Comment (cpcloud) [reply ibis/expr/api.py]: me too, not sure i've developed a stance on it, i try to do what's most clear in the context.

File: ibis/expr/api.py (hunk 6 above)
Comment (kszucs) [reply ibis/expr/api.py]: Crossed out because of this comment.

File: ibis/expr/api.py [hunk 7]
```diff
@@ -2016,21 +2016,34 @@ def _string_dunder_contains(arg, substr):
...
+                )
+            if start is None:
+                start = 0
+
+        if not isinstance(stop, ir.Expr):
+            if stop is not None and stop < 0:
+                raise ValueError(
+                    'Negative slicing not yet supported, got stop value of '
+                    '{:d}'.format(stop)
+                )
```
Comment (kszucs) [inline ibis/expr/api.py]:
```python
if start < 0:
    start = self.length() + start

...

if stop < 0:
    stop = self.length() + stop
```

Comment (cpcloud) [reply ibis/expr/api.py]: Sure, I'm suggesting we do that in a follow up PR.

Comment (kszucs) [reply ibis/expr/api.py]: Ok (however the two lines change is tempting :)

**Comment by cpcloud** [comment]
Merging on green

**Pull request closed without merging**